prog/test/ha_postgres_resource.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+require_relative "../../lib/util"
+
+class Prog::Test::HaPostgresResource < Prog::Test::Base
+  semaphore :destroy
+
+  def self.assemble
+    postgres_service_project = Project.create(name: "Postgres-Service-Project") { _1.id = Config.postgres_service_project_id }
+    postgres_test_project = Project.create(name: "Postgres-Test-Project")
+
+    Strand.create_with_id(
+      prog: "Test::HaPostgresResource",
+      label: "start",
+      stack: [{
+        "postgres_service_project_id" => postgres_service_project.id,
+        "postgres_test_project_id" => postgres_test_project.id
+      }]
+    )
+  end
+
+  label def start
+    st = Prog::Postgres::PostgresResourceNexus.assemble(
+      project_id: frame["postgres_test_project_id"],
+      location: "hetzner-fsn1",
+      name: "postgres-test-standard",
+      target_vm_size: "standard-2",
+      target_storage_size_gib: 128,
+      ha_type: "async"
+    )
+
+    update_stack({"postgres_resource_id" => st.id})
+    hop_wait_postgres_resource
+  end
+
+  label def wait_postgres_resource
+    if postgres_resource.strand.label == "wait" &&
+        postgres_server.run_query("SELECT 1") == "1"
+      hop_test_postgres
+    else
+      nap 10
+    end
+  end
+
+  label def test_postgres
+    unless postgres_server.run_query(test_queries_sql) == "DROP TABLE\nCREATE TABLE\nINSERT 0 10\n4159.90\n415.99\n4.1"
+      update_stack({"fail_message" => "Failed to run test queries"})
+      hop_destroy_postgres
+    end
+
+    hop_trigger_failover
+  end
+
+  label def trigger_failover
+    primary = postgres_resource.servers.find { _1.timeline_access == "push" }
+    update_stack({"primary_ubid" => primary.ubid})
+
+    primary.vm.sshable.cmd("echo -e '\nfoobar' | sudo tee -a /etc/postgresql/#{postgres_resource.version}/main/conf.d/001-service.conf")
+
+    # Get postgres pid and send SIGKILL
+    primary.vm.sshable.cmd("ps aux | grep -v grep | grep '/usr/lib/postgresql/#{postgres_resource.version}/bin/postgres' | awk '{print $2}' | xargs sudo kill -9")
+
+    hop_wait_failover
+  end
+
+  label def wait_failover
+    nap 10 unless postgres_resource.servers.find { _1.timeline_access == "push" }.ubid != frame["primary_ubid"]
+
+    hop_test_postgres_after_failover
+  end
```

Comment:
Not sure if doing this is the best way to test the failover behavior.
I guess external contract would be something like "failovers are expected to take less than x seconds", so should we just wait upto that time and check if a failover happened, and fail if not (instead of waiting until a failover happens).

Reply:
Agreed, normally, it should takes less than 30 seconds. For the tests sake, let's wait 3 minutes and check if failover is completed.